### PR TITLE
Update CMake configuration for easier maintenance

### PR DESCRIPTION
**Pull request by Jettford**
This pull request improves our CMake build system to avoid issues such as everything being inside of a single CMakeLists file and use of things such as glob to gather file names.

### CODE REVIEW

File: dCommon/ZCompression.h [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 #include <cstdint>
 
+#ifndef _WIN32
```
Comment (codeshaunted) [inline dCommon/ZCompression.h:5]: These platform macros are sketchy at best and I think we should move away from using them. Maybe alongside these CMake changes we could introduce platform detection within CMake that defines our own macros (e.g. `DARKFLAME_PLATFORM_WINDOWS`, `DARKFLAME_PLATFORM_LINUX`, etc). CMake can much more reliably detect the platform where combinations of these macros will often fail due to compiler parity issues.

File: dChatFilter/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+set(DCHATFILTER_SOURCES "dChatFilter.cpp")
+
+add_library(dChatFilter ${DCHATFILTER_SOURCES})
```
Comment (codeshaunted) [inline dChatFilter/CMakeLists.txt:3]: We should probably explicitly specify `STATIC` or `SHARED` here if we don't want this to be influenced by the value `BUILD_SHARED_LIBS`. Traditionally DLU has been built with almost entirely static linking, but we could change this as there are several shared dependencies between the server executables.

File: dCommon/ZCompression.h (hunk 1 above)
Comment (aronwk-aaron) [reply dCommon/ZCompression.h:5]: If these weren't reliable, then I think every piece of c++ software would be having issues, no reason to fix what ain't broken

Comment (codeshaunted) [reply dCommon/ZCompression.h:5]: > If these weren't reliable, then I think every piece of c++ software would be having issues, no reason to fix what ain't broken

I recall having issues with clang not working on the initial DLU release due to these macros being used. We also tend to use these platform macros as an analog for compiler (Windows == MSVC, Linux == GCC, etc). It really isn't that hard to implement detection in CMake and it looks nicer in code as well. Here's an example: https://github.com/codeshaunted/soul/blob/0286b126ff89d2e6ba5b285b631e6b1ece28d40f/CMakeLists.txt#L27

File: dChatFilter/CMakeLists.txt (hunk 2 above)
Comment (EmosewaMC) [reply dChatFilter/CMakeLists.txt:3]: Agreed.  Let's not leave anything up to randomness and this is a good chance to reduce the binary size a touch for the libs used across all the binaries like raknet, dCommon, dNet etc.

File: dCommon/ZCompression.h (hunk 1 above)
Comment (Jettford) [reply dCommon/ZCompression.h:5]: Someone else can do this later if wanted, but I won't be.

File: dChatFilter/CMakeLists.txt (hunk 2 above)
Comment (Jettford) [reply dChatFilter/CMakeLists.txt:3]: Yeah that is a good idea. Will do.

**Comment by Jettford** [comment]
Yeah I will revert this to a draft
Too broken at the moment

**Comment by Jettford** [comment]
I can't... seem to revert it to a draft?

**Comment by Jettford** [comment]
Thank you, could not find that.

**Comment by EmosewaMC** [comment]
> WorldServer and ChatServer do not start and seg fault when started. I don't know whats wrong at the moment but this cannot be merged as is.

The issue looks to be something with regex assignment of a regex string.  Investigating...

**Comment by luxaritas** [comment]
Looks like the segfault issue was that the compiler flags were being set after the add_subdirectory calls rather than before, so the compilation was being done with the wrong flags. Moving that seems to have fixed it

File: CMakeLists.txt [hunk 3]
```diff
@@ -177,15 +177,38 @@ file(
...
+# Compiler flags:
+# Disabled deprecated warnings as the MySQL includes have deprecated code in them. 
+# Disabled misleading indentation as DL_LinkedList from RakNet has a weird indent.
+# Disabled no-register
+# Disabled unknown pragmas because Linux doesn't understand Windows pragmas.
+if(UNIX)
+	if(APPLE)
+		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++17 -O2 -Wuninitialized -D_GLIBCXX_USE_CXX11_ABI=0 -D_GLIBCXX_USE_CXX17_ABI=0 -fPIC")
+	else()
+		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++17 -O2 -Wuninitialized -D_GLIBCXX_USE_CXX11_ABI=0 -D_GLIBCXX_USE_CXX17_ABI=0 -static-libgcc -fPIC")
```
Comment (codeshaunted) [inline CMakeLists.txt:189]: `-std=gnu++17` is redundant as we already set the standard to C++17 at the beginning of this file 
https://github.com/DarkflameUniverse/DarkflameServer/blob/17b8e7045f9a0f297e53a721ac3953e71143d9f7/CMakeLists.txt#L5
We should also do 
```cmake
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
in order to enforce the C++17 requirement.

Comment (Jettford) [reply CMakeLists.txt:189]: There are differences between `gnu++17` and `std++17`, there is a requirement for it

Comment (codeshaunted) [reply CMakeLists.txt]: > There are differences between `gnu++17` and `std++17`, there is a requirement for it

The [CMake docs](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html) seem to imply that it should be doing `gnu++`, but alright.

Comment (Jettford) [reply CMakeLists.txt]: The standard is not required however.

**Comment by Jettford** [comment]
@EmosewaMC @aronwk-aaron Could you please test and review, as far as I have tested. It works on Windows and Ubuntu 20

File: CMakeLists.txt (hunk 3 above)
Comment (codeshaunted) [reply CMakeLists.txt]: > The standard is not required however.

I recall C++17 being required for something related to Mersenne Twister... If not, I know it at least C++11 is required, so maybe we should change the general standard requirement to reflect that.

Comment (Jettford) [reply CMakeLists.txt]: I don't think it really matters.

**Comment by EmosewaMC** [comment]
Will test on aarch and mac today

**Review by aronwk-aaron — APPROVED** [review]
LGTM, tested on on ubuntu 18.04

**Review by EmosewaMC — APPROVED** [review]

**Pull request merged**